xrtl/examples/triangle_example.cc

```diff
@@ -0,0 +1,352 @@
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "xrtl/base/logging.h"
+#include "xrtl/base/threading/event.h"
+#include "xrtl/base/threading/thread.h"
+#include "xrtl/gfx/context.h"
+#include "xrtl/gfx/context_factory.h"
+#include "xrtl/testing/demo_main.h"
+#include "xrtl/ui/window.h"
+
+namespace xrtl {
+namespace examples {
+namespace {
+
+using gfx::Buffer;
+using gfx::Context;
+using gfx::ContextFactory;
+using gfx::Framebuffer;
+using gfx::Image;
+using gfx::ImageView;
+using gfx::MemoryPool;
+using gfx::Pipeline;
+using gfx::PipelineLayout;
+using gfx::RenderPass;
+using gfx::RenderPipeline;
+using gfx::RenderState;
+using gfx::ShaderModule;
+using gfx::SwapChain;
+using ui::Control;
+using ui::Window;
+
+class TriangleExample : private Control::Listener {
+ public:
+  TriangleExample() {
+    message_loop_ = MessageLoop::Create();
+    done_event_ = Event::CreateFence();
+  }
+
+  ~TriangleExample() { Thread::Wait(message_loop_->Exit()); }
+
+  ref_ptr<WaitHandle> Run() {
+    // Create and open window.
+    window_ = Window::Create(message_loop_);
+    window_->set_title("Window Demo");
+    auto control = window_->root_control();
+    control->set_listener(this);
+    control->set_size({640, 480});
+    control->set_background_color({255, 0, 0, 255});
```

Comment:
gfx should draw over it - if it doesn't it's a bug, which is why red is useful: you see red, you know you messed up :)